upup/pkg/fi/cloudup/openstack/router.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstack
+
+import (
+	"fmt"
+
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/routers"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kops/util/pkg/vfs"
+)
+
+func (c *openstackCloud) ListRouters(opt routers.ListOpts) ([]routers.Router, error) {
+	var rs []routers.Router
+
+	done, err := vfs.RetryWithBackoff(readBackoff, func() (bool, error) {
+		allPages, err := routers.List(c.neutronClient, opt).AllPages()
+		if err != nil {
+			return false, fmt.Errorf("error listing routers: %v", err)
+		}
+
+		r, err := routers.ExtractRouters(allPages)
+		if err != nil {
+			return false, fmt.Errorf("error extracting routers from pages: %v", err)
+		}
+		rs = r
+		return true, nil
+	})
+	if err != nil {
+		return rs, err
+	} else if done {
+		return rs, nil
+	} else {
+		return rs, wait.ErrWaitTimeout
+	}
+}
+
+func (c *openstackCloud) CreateRouter(opt routers.CreateOptsBuilder) (*routers.Router, error) {
+	var r *routers.Router
+
+	done, err := vfs.RetryWithBackoff(writeBackoff, func() (bool, error) {
+		v, err := routers.Create(c.neutronClient, opt).Extract()
+		if err != nil {
+			return false, fmt.Errorf("error creating router: %v", err)
+		}
+		r = v
+		return true, nil
+	})
+	if err != nil {
+		return r, err
+	} else if done {
+		return r, nil
+	} else {
+		return r, wait.ErrWaitTimeout
+	}
+}
+
+func (c *openstackCloud) CreateRouterInterface(routerID string, opt routers.AddInterfaceOptsBuilder) (*routers.InterfaceInfo, error) {
+	var i *routers.InterfaceInfo
+
+	done, err := vfs.RetryWithBackoff(writeBackoff, func() (bool, error) {
+		v, err := routers.AddInterface(c.neutronClient, routerID, opt).Extract()
+		if err != nil {
```

Comment:
we should ignore error here if there is interface already added to router.

I am currently testing what is working and what is not. After cluster is successfully created, I cannot run update against it anymore because I will get error:
```
I0119 23:52:06.880363   45719 context.go:231] hit maximum retries 5 with error error creating router interface: Bad request with: [PUT https://api.company.com:13696/v2.0/routers/72e64330-acfb-43f7-9957-bfd2797df7f0/add_router_interface], error message: {"NeutronError": {"message": "Bad router request: Router already has a port on subnet e0925881-8edc-45d8-91cc-88b2d9863f70.", "type": "BadRequest", "detail": ""}}
```

Reply:
@zetaab I believe we have not done any work in regards to update afaik. @drekle can correct me though.

Reply:
the fix for this could be something like this https://github.com/zetaab/kops/commit/5d595f731371cb4e7589795c8e7d592b058072b5

it is little bit dirty, but I do not see huge benefit of parsing that json by using "official" way

Reply:
yep the problem is that it does not find the routerinterface correctly. I debugged it little bit, at least the `ID` is empty when querying things (it should be portid i think).

Reply:
@zetaab Mine is found:
```
I0120 00:30:17.107109   28791 executor.go:103] Tasks: 0 done / 87 total; 41 can run
I0120 00:30:19.359116   28791 executor.go:103] Tasks: 41 done / 87 total; 21 can run
I0120 00:30:21.001509   28791 executor.go:103] Tasks: 62 done / 87 total; 11 can run
I0120 00:30:21.679578   28791 executor.go:103] Tasks: 73 done / 87 total; 5 can run
I0120 00:30:22.399036   28791 executor.go:103] Tasks: 78 done / 87 total; 1 can run
I0120 00:30:24.148668   28791 executor.go:103] Tasks: 79 done / 87 total; 4 can run
I0120 00:30:45.574463   28791 context.go:231] hit maximum retries 4 with error GetFloatingIP: fetching floating IP failed: Resource not found
I0120 00:31:40.600914   28791 executor.go:103] Tasks: 82 done / 87 total; 5 can run
I0120 00:32:31.521467   28791 executor.go:103] Tasks: 87 done / 87 total; 0 can run
I0120 00:32:32.923536   28791 update_cluster.go:290] Exporting kubecfg for cluster
```
I do see an issue getting a floating IP by id however in my environment, which doesn't seem to cause issue for me.